scrapi/harvesters/noaa_nodc.py

```diff
@@ -0,0 +1,226 @@
+from datetime import date, timedelta
+
+from lxml import etree
+from nameparser import HumanName
+from requests import exceptions
+
+from scrapi import requests
+from scrapi import settings
+from scrapi.base import XMLHarvester
+from scrapi.base.helpers import compose, single_result, date_formatter, language_codes, coerce_to_list, xml_text_only_list, xml_text_only
+from scrapi.linter.document import RawDocument
+from scrapi.util import copy_to_unicode
+
+logger = logging.getLogger(__name__)
+
+
+def filter_to_publishers(parties):
+    '''Reduce list of ResponsibleParty elements to just publishers'''
+    return filter_responsible_parties(parties, ['publisher'])
+
+
+def filter_to_contributors(parties):
+    '''Reduce list of ResponsibleParty elements to just contributors'''
+    return filter_responsible_parties(parties, ['resourceProvider'])
+
+
+def filter_responsible_parties(parties, roles):
+    '''Return list of ResponsibleParties whose role is in roles param'''
+    return [party for party in parties if party.xpath('./gmd:role/gmd:CI_RoleCode/node()', namespaces=party.nsmap)[0] in roles]
+
+
+def parse_contributors(parties):
+    '''Turn list of gmd:CI_ResponsibleParty elements into
+    SHARE-compliant contributors list
+    '''
+
+    # The NODC schema doesn't explicitly distinguish between
+    # organizations and individuals.  Every party has an
+    # organizationName tag, individuals have an individualName tag as
+    # well.  The email address is tied to the party, so we can
+    contributors = []
+    for party in parties:
+        contributor = {}
+        individual = extract_individual(party)
+        organization = extract_organization(party)
+        if individual:
+            contributor = individual
+            if organization:
+                contributor['affiliation'] = [organization]
+        elif organization:
+            contributor = organization
+        else:
+            raise NotImplementedError
+
+        # the email address is tied to the party, so wait until we know
+        # who the contributor is before extracting
+        email = party.xpath('./gmd:contactInfo/gmd:CI_Contact/gmd:address/gmd:CI_Address/gmd:electronicMailAddress', namespaces=party.nsmap)
+        if len(email) > 0:
+            contributor['email'] = xml_text_only(email[0])
+        contributors.append(contributor)
+
+    return contributors
+
+
+def extract_individual(party):
+    '''Return SHARE-compliant person object from a CI_ResponsibleParty element'''
+    individual = party.xpath('./gmd:individualName/gmx:Anchor', namespaces=party.nsmap)
+    if len(individual) > 0:
+        name = HumanName(individual[0].text)
+        ns_prefix = party.nsmap['xlink']
+        sameAs = individual[0].get('{' + ns_prefix + '}href')
+        return {
+            'name': individual[0].text,
+            'givenName': name.first,
+            'additionalName': name.middle,
+            'familyName': name.last,
+            'sameAs': [sameAs],
```

Comment:
I think if there is nothing to fill the sameas field, this will be ```[None]```, which is sort of messy. I'd prefer that you conditionally add the sameAs field based on whether or not there is something to put in it.